File: examples/python/10_persistence_lite/agent.py
"""
Persistence Lite — pluggable TaskStore + PushNotifier example.

Shows how to:
  1. Swap the protocol-level TaskStore (e.g. use Redis/Postgres in production
     by replacing InMemoryTaskStore with your own implementation).
  2. Use LogPushNotifier during development (zero config).
  3. Switch to WebhookPushNotifier for production by setting WEBHOOK_URL (and
     optionally WEBHOOK_SECRET) environment variables.
  4. Wire both into an Agent with only a few extra lines.

Run (dev mode, no env vars needed):
    python agent.py

Run (production mode with webhook):
    WEBHOOK_URL=https://hooks.example.com/a2a python agent.py

Optionally sign payloads:
    WEBHOOK_URL=https://hooks.example.com/a2a WEBHOOK_SECRET=mysecret python agent.py
"""

from __future__ import annotations

import asyncio
import os

from a2a.server.tasks import InMemoryTaskStore

from a2a_lite import Agent
from a2a_lite.push_notifications import LogPushNotifier, WebhookPushNotifier

# ---------------------------------------------------------------------------
# Push notifier — dev vs production
# ---------------------------------------------------------------------------

WEBHOOK_URL = os.getenv("WEBHOOK_URL")
WEBHOOK_SECRET = os.getenv("WEBHOOK_SECRET")  # optional

if WEBHOOK_URL:
    # Production: POST completed-task events to a real webhook endpoint.
    # Set WEBHOOK_SECRET to have each request signed with HMAC-SHA256.
    push_notifier = WebhookPushNotifier(
        url=WEBHOOK_URL,
        secret=WEBHOOK_SECRET,  # None → no signature header
        timeout=5.0,
        max_retries=2,
    )
else:
    # Development: just log events — no external service needed.
    push_notifier = LogPushNotifier()

# ---------------------------------------------------------------------------
# Protocol-level TaskStore
# ---------------------------------------------------------------------------
# In a real application you would swap InMemoryTaskStore() for a persistent
# implementation (e.g. a Redis-backed store).  Passing it explicitly here
# makes the seam visible and easy to test.

protocol_store = InMemoryTaskStore()

# ---------------------------------------------------------------------------
# Agent
# ---------------------------------------------------------------------------

agent = Agent(
    name="PersistenceLiteAgent",
    description=(
        "Demonstrates pluggable TaskStore and PushNotifier. "
        "Uses LogPushNotifier in dev and WebhookPushNotifier when "
        "WEBHOOK_URL is set."
    ),
    version="1.0.0",
    # Hand the explicit store to the protocol layer.
    protocol_task_store=protocol_store,
    # Every completed skill invocation triggers the notifier.
    push_notifier=push_notifier,
)


# ---------------------------------------------------------------------------
# Skills
# ---------------------------------------------------------------------------


@agent.skill("echo", description="Return the input message unchanged.")
async def echo(message: str) -> dict:
    """Echo the caller's message back."""
    return {"echo": message}


@agent.skill(
    "slow_sum",
    description=(
        "Add two integers with a simulated delay. "
        "Demonstrates a long-running skill alongside push notifications."
    ),
)
async def slow_sum(a: int, b: int, delay: float = 1.0) -> dict:
    """Add a + b after sleeping for *delay* seconds (default 1 s)."""
    await asyncio.sleep(delay)
    return {
        "a": a,
        "b": b,
        "sum": a + b,
        "delay_seconds": delay,
    }


# ---------------------------------------------------------------------------
# Entry point
# ---------------------------------------------------------------------------

if __name__ == "__main__":
    notifier_type = "WebhookPushNotifier" if WEBHOOK_URL else "LogPushNotifier"
    print(f"Push notifier : {notifier_type}")
    if WEBHOOK_URL:
        print(f"Webhook URL   : {WEBHOOK_URL}")
        print(f"Signed        : {'yes' if WEBHOOK_SECRET else 'no'}")
    print(f"TaskStore     : {type(protocol_store).__name__}")
    print()

    agent.run(port=8793)

File: examples/typescript/10_persistence_lite/agent.ts
/**
 * Persistence Lite — pluggable TaskStore + PushNotifier example.
 *
 * Demonstrates:
 *   1. Passing an explicit InMemoryTaskStore as protocolTaskStore so the SDK's
 *      request handler uses YOUR store instead of creating its own.
 *   2. Selecting LogPushNotifier (dev) or WebhookPushNotifier (prod) based on
 *      environment variables.
 *   3. Two skills: a fast `echo` and a slow `slowSum` with a deliberate delay.
 *
 * Run (dev):
 *   npx tsx examples/typescript/10_persistence_lite/agent.ts
 *
 * Run (production webhook):
 *   WEBHOOK_URL=https://your.server/hook \
 *   WEBHOOK_SECRET=my-signing-secret \
 *   npx tsx examples/typescript/10_persistence_lite/agent.ts
 */

import { Agent, InMemoryTaskStore, LogPushNotifier, WebhookPushNotifier } from 'a2a-lite';
import type { PushNotifier } from 'a2a-lite';
import { InMemoryTaskStore as SdkTaskStore } from '@a2a-js/sdk/server';

// ---------------------------------------------------------------------------
// 1. Task store — shared instance you own and can inspect at runtime
// ---------------------------------------------------------------------------

// This is the lite TaskStore used for TaskContext injection within skills.
const liteStore = new InMemoryTaskStore();

// This is the SDK-level TaskStore used by the A2A protocol layer (request
// deduplication, task lifecycle).  By passing it explicitly you can swap it
// for a Redis or Postgres-backed implementation without changing anything else.
const protocolStore = new SdkTaskStore();

// ---------------------------------------------------------------------------
// 2. Push notifier — dev vs. production
// ---------------------------------------------------------------------------

function buildPushNotifier(): PushNotifier {
  const webhookUrl = process.env.WEBHOOK_URL;

  if (webhookUrl) {
    const secret = process.env.WEBHOOK_SECRET;
    console.log(`[config] Using WebhookPushNotifier → ${webhookUrl}${secret ? ' (signed)' : ''}`);
    return new WebhookPushNotifier({
      url: webhookUrl,
      secret,
      maxRetries: 3, // exponential back-off is built in
      timeoutMs: 10_000,
    });
  }

  console.log('[config] Using LogPushNotifier (set WEBHOOK_URL to switch to webhook mode)');
  return new LogPushNotifier();
}

// ---------------------------------------------------------------------------
// 3. Agent
// ---------------------------------------------------------------------------

const agent = new Agent({
  name: 'PersistenceLiteAgent',
  description: 'Demonstrates pluggable TaskStore and PushNotifier',
  version: '1.0.0',

  // Lite TaskStore — enables TaskContext in skills
  taskStore: liteStore,

  // Protocol TaskStore — passed to the SDK's DefaultRequestHandler
  protocolTaskStore: protocolStore,

  // Push notifier — fires after every skill completion
  pushNotifier: buildPushNotifier(),
});

// ---------------------------------------------------------------------------
// 4. Skills
// ---------------------------------------------------------------------------

/**
 * echo — fast, returns immediately.
 *
 * Call: { "skill": "echo", "params": { "message": "hello" } }
 */
agent.skill('echo', { description: 'Echo a message back' }, async (params) => {
  const { message } = params as { message: string };
  return { echoed: message, at: new Date().toISOString() };
});

/**
 * slowSum — adds two numbers after a deliberate delay.
 * This makes it easy to observe push notifications arriving after a pause.
 *
 * Call: { "skill": "slowSum", "params": { "a": 3, "b": 4, "delayMs": 2000 } }
 */
agent.skill(
  'slowSum',
  { description: 'Add two numbers after a delay (demonstrates async push notification)' },
  async (params) => {
    const { a, b, delayMs = 1000 } = params as { a: number; b: number; delayMs?: number };
    await new Promise((resolve) => setTimeout(resolve, delayMs));
    return { sum: Number(a) + Number(b), delayMs };
  }
);

// ---------------------------------------------------------------------------
// 5. Startup hook — log store contents when the agent stops
// ---------------------------------------------------------------------------

agent.onShutdown(() => {
  const tasks = liteStore.list();
  if (tasks.length > 0) {
    console.log(`\n[shutdown] ${tasks.length} task(s) in lite store at exit:`);
    for (const t of tasks) {
      console.log(`  • ${t.id} [${t.skill}] → ${t.status.state}`);
    }
  }
});

// ---------------------------------------------------------------------------
// 6. Run
// ---------------------------------------------------------------------------

agent.run({ port: 8790 });
